### PR TITLE
enhance: Rewrite gen segment plan based on assign segment

### PR DESCRIPTION
**Pull request by weiliu1031**
issue: #29582
This PR rewrite gen segment plan logic based on assign segment in `score_based_balancer`

### CODE REVIEW

File: internal/querycoordv2/balance/score_based_balancer.go [hunk 1]
```diff
@@ -232,91 +248,65 @@ func (b *ScoreBasedBalancer) genStoppingSegmentPlan(replica *meta.Replica, onlin
...
-				// not provide sufficient balance benefit, we will seize balancing in this round
+		sort.Slice(segments, func(i, j int) bool {
+			return segments[i].GetNumOfRows() < segments[j].GetNumOfRows()
+		})
+		for _, s := range segments {
+			segmentsToMove = append(segmentsToMove, s)
+			leftScore -= b.calculateSegmentScore(s)
+			if leftScore <= average {
 				break
 			}
```
Comment (congqixia) [inline internal/querycoordv2/balance/score_based_balancer.go:292]: these code may case node score below avg score and cause segment bouncing between nodes

Comment (weiliu1031) [reply internal/querycoordv2/balance/score_based_balancer.go:292]: we calculate the benefit during the `assignSegment`, if  reassign won't get enough benefits, we will skip this reassignment

**Review by congqixia — APPROVED** [review]
/lgtm

**Pull request merged**